lib/features/email/data/local/html_analyzer.dart

```diff
@@ -112,35 +122,76 @@ class HtmlAnalyzer {
-
-    final listInlineAttachment = await Future.wait(listImgTag.map((imgTag) async {
-      final idImg = imgTag.attributes['id'];
-      final cid = idImg!.replaceFirst('cid:', '').trim();
-      imgTag.attributes['src'] = 'cid:$cid';
-      imgTag.attributes.remove('id');
-      return cid;
-    })).then((listCid) {
-      final listInlineAttachment = listCid
-        .map((cid) {
-          if (inlineAttachments.containsKey(cid)) {
-            return inlineAttachments[cid]!.toEmailBodyPart(charset: Constant.base64Charset);
-          } else {
-            return null;
-          }
-        })
-        .whereNotNull()
-        .toSet();
+    final listImgTag = document.querySelectorAll('img[src^="data:image/"]');
 
-      return listInlineAttachment;
-    });
+    log('HtmlAnalyzer::replaceImageBase64ToImageCID:listImgTagLength = ${listImgTag.length} | inlineAttachments = ${inlineAttachments.length}');
 
-    final newContent = document.body?.innerHtml ?? emailContent;
+    if (listImgTag.isEmpty) {
+      return Tuple2(
+        emailContent,
+        inlineAttachments.isNotEmpty
+          ? inlineAttachments.values.toList().toEmailBodyPart(charset: Constant.base64Charset)
+          : {},
+      );
+    }
+
+    final Set<EmailBodyPart> inlineAttachmentsSet = {};
+    final List<Future<void>> asyncTasks = [];
+
+    for (final imgTag in listImgTag) {
+      final attributes = imgTag.attributes;
+      final imageSrc = attributes['src'];
+      if (imageSrc?.isEmpty ?? true) continue;
+
+      final idImg = attributes['id'];
+      if (idImg?.startsWith(cidPrefixKey) == true) {
+        final cid = idImg!.substring(cidPrefixKey.length).trim();
+        final attachment = inlineAttachments[cid];
+
+        if (attachment != null) {
+          attributes['src'] = '$cidPrefixKey$cid';
+          attributes.remove('id');
+          inlineAttachmentsSet.add(attachment.toEmailBodyPart(charset: Constant.base64Charset));
+          continue;
+        }
+      }
+
+      if (uploadUri == null) continue;
+
+      final taskId = idImg?.startsWith(cidPrefixKey) == true
+        ? idImg!.substring(cidPrefixKey.length)
+        : _uuid.v1();
+
+      asyncTasks.add(_retrieveAttachmentFromUpload(
+        taskId: taskId,
+        uploadUri: uploadUri,
+        base64ImageTag: imageSrc!,
+      ).then((attachmentRecord) {
+        if (attachmentRecord == null) return;
 
-    return Tuple2(newContent, listInlineAttachment);
+        final newInlineAttachment = attachmentRecord.$1.toAttachmentWithDisposition(
+          disposition: ContentDisposition.inline,
+          cid: attachmentRecord.$2,
+        );
+
+        final newCid = newInlineAttachment.cid!;
+        inlineAttachments[newCid] = newInlineAttachment;
+        attributes['src'] = '$cidPrefixKey$newCid';
+        attributes.remove('id');
+
+        inlineAttachmentsSet.add(newInlineAttachment.toEmailBodyPart(charset: Constant.base64Charset));
```

Comment:
I am worried about much of the `async task`, it may add together in the same `inlineAttachmentSet`. Can we synchronize add it? And please add the test case to make sure it can handle things in parallel.

Reply:
Update unit test for 100 image tag